tpm2/policy_v3.go

```diff
@@ -148,10 +158,61 @@ type staticPolicyData_v3 struct {
 	RequireAuthValue       bool
 }
 
+// policyOrData_v3 is the version 3 on-disk representation of policyOrTree.
+// It is a flattened tree which suitable for serializing - the tree is just
+// a slice of nodes, with each node specifying an offset to its parent node.
+type policyOrData_v3 = policyOrData_v0
+
 // pcrPolicyData_v3 represents version 3 of the PCR policy metadata for
 // executing a policy session, and can be updated. It has the same format
 // as version 2.
-type pcrPolicyData_v3 = pcrPolicyData_v2
+type pcrPolicyData_v3 struct {
+	// This is mostly the same as v0-v2, with the only difference being
+	// how the call to executeRevocationCheck works.
+	pcrPolicyData_v2
+}
+
+func (d *pcrPolicyData_v3) executeRevocationCheck(tpm *tpm2.TPMContext, counter tpm2.ResourceContext, updateKey *tpm2.Public, policySession tpm2.SessionContext) error {
+	operandB := make([]byte, 8)
+	binary.BigEndian.PutUint64(operandB, d.PolicySequence)
+
+	// XXX(chrisccoulson): This is the 3rd session we open here, with all 3 loaded
+	// on the TPM. PC Client TPMs are only guaranteed to support 3 loaded sessions,
+	// so we're pushing things a bit close to the wire here. This code does execute
+	// during early boot, but it might be better to make the HMAC session associated
+	// with the TPM connection unloaded by default (context saved), and only context
+	// load it when needed.
+	revocationCheckSession, err := tpm.StartAuthSession(nil, nil, tpm2.SessionTypePolicy, nil, counter.Name().Algorithm())
+	if err != nil {
+		return fmt.Errorf("cannot create policy session for revocation check: %w", err)
+	}
+	defer tpm.FlushContext(revocationCheckSession)
+
+	authPolicies, err := computeV3PcrPolicyCounterAuthPolicies(counter.Name().Algorithm(), updateKey)
+	if err != nil {
+		return policyDataError{fmt.Errorf("cannot compute auth policies for PCR policy counter: %w", err)}
+	}
+	if err := tpm.PolicyCommandCode(revocationCheckSession, tpm2.CommandPolicyNV); err != nil {
+		return err
+	}
+	if err := tpm.PolicyOR(revocationCheckSession, authPolicies); err != nil {
+		return err
+	}
```

Comment:
it would be good to have a comment here that explain why we create a new session and what we do with it

Reply:
I've added a comment now.